typing/ctype.ml

```diff
@@ -5450,17 +5450,47 @@ let same_constr env t1 t2 =
 let () =
   Env.same_constr := same_constr
 
-let immediacy env typ =
-   match get_desc typ with
-  | Tconstr(p, _args, _abbrev) ->
-    begin try
-      let type_decl = Env.find_type p env in
-      type_decl.type_immediate
-    with Not_found -> Type_immediacy.Unknown
-    (* This can happen due to e.g. missing -I options,
-       causing some .cmi files to be unavailable.
-       Maybe we should emit a warning. *)
+(* We use expand_head_opt version of expand_head to get access
+   to the manifest type of private abbreviations. *)
+let rec get_unboxed_type_representation env ty fuel =
+  if fuel < 0 then ty else
```

Comment:
This commit changes the return type of `get_unboxed_type_representation` to return a type instead of a type option. Before we would return `None` when reaching the fuel limit (but not when an expansion cannot be found), and now we return the type that we have at the point of reaching the limit.

My intuition is that it is better to know that we failed to reach a normal form, so I think that silently returning a non-normal type is not ideal. On the other hand, returning the last non-normal type reached is more informative than `None`. I think that a perfect return type for this function would be `(ty, ty) result`, using `Error ty` here. This also gives us the option to fail with an error if we don't reach the normal form, which is possibly what should happen anyway -- it is arguably a mistake to allow programs with infinite cycles modulo unboxing, as in https://github.com/ocaml/ocaml/issues/7424.

I would be in favor of sticking with `ty option` or going with `(ty, ty) result` instead of moving to `ty`, to avoid making the interface less general. This makes the caller code slightly less convenient, but we could have a derived `get_unboxed_type_approximation` or something that returns just a `ty` for convenience.

This being said, I am hoping to get rid of fuel and this failure mode completely in the medium-term future by reviving https://github.com/ocaml/ocaml/pull/10479 .

Reply:
Sorry, I should not have slipped this somewhat orthogonal change in.  I do rather like the current version, since nobody is using the failure information, but I'm happy with the `result` compromise and will make this change tomorrow.

Reply:
I have now changed this to the `result` version.